tree/forest/v7/inc/ROOT/RTreeStorage.hxx

```diff
@@ -0,0 +1,112 @@
+
+namespace Detail {
+
+class RBranchBase;
+class RColumn;
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RTreeStorage
+\ingroup Forest
+\brief Manages tree meta-data, which is common for sinks and sources.
+
+The tree meta-data contains of a list of branches, a unique identifier, and provenance information.
+*/
+// clang-format on
+class RTreeStorage {
+   // TODO(jblomer)
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RTreeSink
+\ingroup Forest
+\brief Abstract interface to write data into a tree
+
+The tree sink takes the list of columns and afterwards a series of page commits and cluster commits.
+The user is responsible to commit clusters at consistent point, i.e. when all pages corresponding to data
+up to the given entry number are committed.
+*/
+// clang-format on
+class RTreeSink : public RTreeStorage {
+public:
+   RTreeSink(std::string_view treeName);
+   virtual ~RTreeSink();
+
+   /// Physically creates the storage container to hold the tree (e.g., a directory in a TFile or a S3 bucket)
+   virtual void Create() = 0;
+   /// Register a new column
+   virtual void AddColumn(RColumn *column) = 0;
+   /// Write a page to the storage. The column attached to the page must have been added before.
+   virtual void CommitPage(RPage *page) = 0;
+   /// Finalize the current cluster and create a new one for the following data.
+   virtual void CommitCluster(TreeIndex_t nEntries) = 0;
+   /// Finalize the current cluster and the entrire data set.
+   virtual void CommitDataset(TreeIndex_t nEntries) = 0;
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RTreeSource
+\ingroup Forest
+\brief Abstract interface to read data from a tree
+
+The tree source is initialized with the columns of interest. Pages from those columns can then be
+mapped into pages. The tree source also gives access to its meta-data.
+*/
+// clang-format on
+class RTreeSource : public RTreeStorage {
+public:
+   RTreeSource(std::string_view treeName);
+   virtual ~RTreeSource();
+   /// TODO: copy/assignment for creating clones in multiple threads.
+
+   /// Open the physical storage container for the tree
+   virtual void Attach() {/* Make me abstract */}
+   /// Register a column for reading values. The column must exist in the tree on disk corresponding to the meta-data
+   /// provided by the column object.
+   virtual void AddColumn(RColumn * /*column*/) {/* Make me abstract */}
+   /// Return a top-level branch that can be iterated and contains as children all the branches stored in the tree.
+
+   // TODO(jblomer): virtual std::unique_ptr<RBranchBase> ListBranches() {/* Make me abstract */ return nullptr;}
+   // TODO(jblomer): ListClusters()
+
+   /// Fills a page starting with index rangeStart; the corresponding column is taken from the page object
+   void MapSlice(TreeIndex_t /*rangeStart*/, RPage * /*page*/) {/* Make me abstract */}
+   virtual TreeIndex_t GetNentries() {/* Make me abstract */ return 0;}
```

Comment:
Number of entries looks odd being returned as `TreeIndex_t`.

Reply:
[note: naming conventions says GetNEntries but the legacy equivalent is GetEntries.]

Reply:
`GetEntries` is confusing as `GetEntry` gets an entry, and one could assume that `GetEntries` would then get all entries in a container, not the number of entries.